[Paywall Experiments] Implement ExPlat

mebarbosa opened the pull request:
## Description
- The goal of this PR is to replace the AB test control for the paywall experiments from Firebase to Explat
- We are going to keep this behind feature flag: p1726143653353199/1725977109.174469-slack-C07G1MH890E 
- To be able to use Explat you need to have `EXPLAT_EXPERIMENT` feature flag enabled
- See for more context: paqN3M-1fz-p2#comment-1310
- See the experiment: 21899-explat-experiment
- See ExPlat SDK: https://github.com/Automattic/Automattic-Tracks-Android/tree/trunk/experimentation#readme
- See iOS PR: https://github.com/Automattic/pocket-casts-ios/pull/1969


Fixes #2870

## Testing Instructions
- Mandatory to have the `EXPLAT_EXPERIMENT` FF enabled

I am currently trying to figure out how to change from Control to Treatment to see both flows for the AB testing. You can see the slack thread here: p1726145194421799-slack-C07DLM97HRQ


## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
- [ ] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

dangermattic commented:
<!--
  1 Error: PR requires a [Type] label and...
  1 Warning: Class <code>ExperimentLogger</...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires a [Type] label and either a [Area] or [Project] label.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ExperimentLogger</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mebarbosa commented on modules/features/account/src/main/java/au/com/shiftyjelly/pocketcasts/account/viewmodel/OnboardingUpgradeFeaturesViewModel.kt:
```diff
@@ -103,14 +103,12 @@ class OnboardingUpgradeFeaturesViewModel @Inject constructor(
                     UpgradeLayout.Original
                 } else {
-                    UpgradeLayout.Original
+                    UpgradeLayout.Reviews
```
I am setting this to `Reviews` for testing purposes, but we might want to change it to `UpgradeLayout.Original` so we can run an AA test for Explat

necronet replied:
Probably since it's an A/A Explat we want both variation to have the same experience and only test that the assignment process is working properly

mebarbosa replied:
Yep! I was planning to revert this before merge this PR. It's done now! Thanks for checking this

mebarbosa commented on modules/services/analytics/src/main/java/au/com/shiftyjelly/pocketcasts/analytics/experiments/Experiment.kt:
```diff
@@ -1,5 +1,10 @@
-sealed class Experiment(val identifier: String) {
-    data object PaywallAATest : Experiment(identifier = "pocketcasts_paywall_android_aa_test")
+enum class Experiment(val identifier: String) {
+    PaywallAATest("pocketcasts_paywall_android_aa_test"),
```
21899-explat-experiment

wzieba commented on gradle/libs.versions.toml:
```diff
@@ -52,6 +52,7 @@ showkase = "1.0.3"
 tasker-plugin = "0.4.10"
 test = "1.6.1"
 tracks = "5.1.2"
+explat = "6.0.1"
```
```suggestion
tracks = "6.0.1"
```

wzieba commented on gradle/libs.versions.toml:
```diff
@@ -226,6 +227,7 @@ androidx-test-runner = { module = "androidx.test:runner", version.ref = "test" }
 # Tracks
 automattic-crashlogging = { module = "com.automattic.tracks:crashlogging", version.ref = "tracks" }
 automattic-tracks = { module = "com.automattic:Automattic-Tracks-Android", version.ref = "tracks" }
+automattic-explat = { module = "com.automattic.tracks:experimentation", version.ref = "explat" }
```
```suggestion
automattic-explat = { module = "com.automattic.tracks:experimentation", version.ref = "tracks" }
```

All the "tracks" modules are shipped from the same repository and use the same version. To reduce complexity, I propose unifying it.

wzieba reviewed (commented):
Looks good, thank you for working on this @mebarbosa ! I added some suggestions.

MiSikora commented on modules/services/analytics/src/main/java/au/com/shiftyjelly/pocketcasts/analytics/di/ExperimentModule.kt:
```diff
@@ -0,0 +1,64 @@
+        logger: ExperimentLogger,
+        @Cached okHttpClient: OkHttpClient,
+    ): VariationsRepository {
+        val cacheDir = File(context.cacheDir, "experiments_cache").apply {
+            if (!exists()) mkdirs()
+        }
```
@wzieba Do you know if `VariationsRepository` repository handles file creation? I'm asking because we shouldn't be doing IO operations on the main thread nor in the DI graph.

wzieba replied:
At this moment, no, it doesn't. I can add such an addition, though.

But I wonder, wouldn't simple `cacheDir` work fine here?

```suggestion
        val cacheDir = ontext.cacheDir
```

WDYT?

MiSikora replied:
Using `cacheDir` would be fine but this directory can also not exists.

MiSikora replied:
Also, a separate point. But after thinking about it I don't think we should be using `cacheDir` for anything related to experiments. This directory can be easily wiped by the system. We should use something less volatile like `filesDir`.

mebarbosa replied:
Great point about `filesDir`. That makes sense to me. What do you think @wzieba ?

wzieba replied:
Sounds good to me 👍 The ExPlat Android SDK should handle a case with removed cache as well, but it's still good to make maximum use of it (use cache as long as it's valid). Especially, that the cache is so tiny.

Related to this, I'm adding an improvement: https://github.com/Automattic/Automattic-Tracks-Android/pull/248, so you could easily provide cache directory as a subdirectory of `filesDir`. The improvement should be available soon!

wzieba replied:
Tracks `6.0.2` is now available that covers a case of non-existing parent directory 👍

MiSikora commented on modules/services/analytics/src/test/kotlin/au/com/shiftyjelly/pocketcasts/analytics/experiments/ExperimentProviderTest.kt:
```diff
@@ -0,0 +1,95 @@
+
+    @Before
+    fun setUp() {
+        MockitoAnnotations.openMocks(this)
```
Is this call needed?

mebarbosa replied:
No need! Nice catch 👍🏽

mebarbosa commented on modules/services/analytics/src/main/java/au/com/shiftyjelly/pocketcasts/analytics/experiments/ExperimentLogger.kt:
```diff
@@ -0,0 +1,14 @@
+import au.com.shiftyjelly.pocketcasts.utils.log.LogBuffer
+import com.automattic.android.experimentation.ExperimentLogger
+
+class ExperimentLogger : ExperimentLogger {
```
You are right! I've simplified the code and you can check here [15a528a](https://github.com/Automattic/pocket-casts-android/pull/2871/commits/15a528a21405842b0e15ce994d8ebe0582ff2c16)

mebarbosa commented on modules/services/analytics/src/main/java/au/com/shiftyjelly/pocketcasts/analytics/experiments/ExperimentProvider.kt:
```diff
@@ -1,18 +1,52 @@
+import javax.inject.Singleton
+import au.com.shiftyjelly.pocketcasts.analytics.experiments.Experiment as ExperimentModel
 
+@Singleton
```
Got it! Removed here [3b79672](https://github.com/Automattic/pocket-casts-android/pull/2871/commits/3b79672843d6afc2be06bdea171f9dc3685d66c4)

mebarbosa commented on modules/services/analytics/src/main/java/au/com/shiftyjelly/pocketcasts/analytics/experiments/Variation.kt:
```diff
@@ -2,13 +2,5 @@ package au.com.shiftyjelly.pocketcasts.analytics.experiments
-        fun fromName(name: String) = when (name) {
-            TREATMENT -> Treatment
-            else -> Control
-        }
-    }
+    data class Treatment(val name: String? = null) : Variation()
 }
```
I don't have a strong opinion about this. I was considering having this class act as a bridge so that if we ever change the API from ExPlat to another, we would only need to update the code in a few places rather than in all the experiments

wzieba replied:
Sounds good to me to leave it too 👍 I don't predict this api to ever change, but if having this additional layer is preferable, I'm completely ok with it!

wzieba commented on modules/services/analytics/src/main/java/au/com/shiftyjelly/pocketcasts/analytics/di/ExperimentModule.kt:
```diff
@@ -0,0 +1,68 @@
+            platform = ExperimentProvider.PLATFORM,
+            experiments = experiments,
+            logger = logger,
+            failFast = true,
```
```suggestion
            failFast = BuildConfig.DEBUG,
```

Can we please fail fast only in debug? This was intended like this - I think we shouldn't crash the app for users if a code related to experiments is failing.

mebarbosa replied:
Nice catch! I will update to this

wzieba commented on modules/services/analytics/src/main/java/au/com/shiftyjelly/pocketcasts/analytics/di/ExperimentModule.kt:
```diff
@@ -0,0 +1,69 @@
+        @ApplicationContext context: Context,
+        @Cached okHttpClient: OkHttpClient,
+    ): VariationsRepository {
+        val directory = File(context.filesDir, "experiments").apply {
+            if (!exists()) mkdirs()
+        }
```
```suggestion
        val directory = File(context.filesDir, "experiments")
```

The Tracks/ExPlat SDK will handle the case of missing directory now (just to double-check, it'd be great to verify if it's true on fresh app installation 🙏 )

necronet reviewed (commented):
Overall looks good just one comment maybe a blocker which is the feature flag when using the `initialize` function in the application class!

mebarbosa commented on modules/features/account/src/main/java/au/com/shiftyjelly/pocketcasts/account/viewmodel/OnboardingLogInViewModel.kt:
```diff
@@ -77,6 +79,7 @@ class OnboardingLogInViewModel @Inject constructor(
             when (result) {
                 is LoginResult.Success -> {
                     podcastManager.refreshPodcastsAfterSignIn()
+                    experiments.initialize()
```
Yes! I wrapped all calls behind this feature flag [10a3d82](https://github.com/Automattic/pocket-casts-android/pull/2871/commits/10a3d82039dadd1d639aaea6d2c87b06a807cd8b)

mebarbosa commented:
@necronet @wzieba I addressed all feedbacks and suggestions. Would you mind to take a look at the PR again so we can close it in case there is nothing more to address here? 🙏

necronet reviewed (approved):
LGTM. ExPlat Approved

wzieba reviewed (approved):
Great work!
